db/db_impl/db_impl_write.cc

```diff
@@ -1737,6 +1737,178 @@ void DBImpl::NotifyOnMemTableSealed(ColumnFamilyData* /*cfd*/,
 }
 #endif  // ROCKSDB_LITE
 
+Status DBImpl::MemPurge(ColumnFamilyData* cfd, MemTable* new_mem) {
+  Status s;
+  assert(new_mem != nullptr);
+
+  JobContext job_context(next_job_id_.fetch_add(1), true);
+  std::vector<SequenceNumber> snapshot_seqs;
+  SequenceNumber earliest_write_conflict_snapshot;
+  SnapshotChecker* snapshot_checker;
+  GetSnapshotContext(&job_context, &snapshot_seqs,
+                     &earliest_write_conflict_snapshot, &snapshot_checker);
+
+  // Grab current memtable
+  MemTable* m = cfd->mem();
+  SequenceNumber earliest_seqno = m->GetEarliestSequenceNumber();
+
+  // Create two iterators, one for the memtable data (contains
+  // info from puts + deletes), and one for the memtable
+  // Range Tombstones (from DeleteRanges).
+  ReadOptions ro;
+  ro.total_order_seek = true;
+  Arena arena;
+  std::vector<InternalIterator*> memtables(1, m->NewIterator(ro, &arena));
+  std::vector<std::unique_ptr<FragmentedRangeTombstoneIterator>>
+      range_del_iters;
+  auto* range_del_iter = m->NewRangeTombstoneIterator(ro, kMaxSequenceNumber);
+  if (range_del_iter != nullptr) {
+    range_del_iters.emplace_back(range_del_iter);
+  }
+  ScopedArenaIterator iter(
+      NewMergingIterator(&(cfd->internal_comparator()), memtables.data(),
+                         static_cast<int>(memtables.size()), &arena));
+
+  auto* ioptions = cfd->ioptions();
+
+  // Place iterator at the First (meaning most recent) key node.
+  iter->SeekToFirst();
+
+  std::unique_ptr<CompactionRangeDelAggregator> range_del_agg(
+      new CompactionRangeDelAggregator(&(cfd->internal_comparator()),
+                                       snapshot_seqs));
+  for (auto& rd_iter : range_del_iters) {
+    range_del_agg->AddTombstones(std::move(rd_iter));
+  }
+
+  // If there is valid data in the memtable,
+  // or at least range tombstones, copy over the info
+  // to the new memtable.
+  if (iter->Valid() || !range_del_agg->IsEmpty()) {
+    std::unique_ptr<CompactionFilter> compaction_filter;
+    if (ioptions->compaction_filter_factory != nullptr &&
+        ioptions->compaction_filter_factory->ShouldFilterTableFileCreation(
+            TableFileCreationReason::kFlush)) {
+      CompactionFilter::Context ctx;
+      ctx.is_full_compaction = false;
+      ctx.is_manual_compaction = false;
+      ctx.column_family_id = cfd->GetID();
+      ctx.reason = TableFileCreationReason::kFlush;
+      compaction_filter =
+          ioptions->compaction_filter_factory->CreateCompactionFilter(ctx);
+      if (compaction_filter != nullptr &&
+          !compaction_filter->IgnoreSnapshots()) {
+        s.PermitUncheckedError();
```

Comment:
you can assign the ```NotSupported``` status to ```s``` and ```return s```. That way you don't need to add ```s.PermitUncheckedError();```. PermitUncheckedError is added if we are sure the status is not checked.